File: leetcode-js/159-longest-substring-with-at-most-two-distinct-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstringTwoDistinct = function(s) {
    // left and right boundary of the substring
    let left = 0;
    let right = -1;

    let maxLen = 0;

    for (let i = 1; i < s.length; i += 1) {
        if (s[i-1] !== s[i]) {
            // When current char doesn't equal to last char,
            // check whether current char equal to the right boundary char
            if (right >= 0 && s[i] !== s[right]) {
                // if current char not equal to right boundary char,
                // update max length
                // and update left boundary to shortern the substring
                maxLen = Math.max(maxLen, i - left);
                left = right + 1;
            }

            right = i - 1;
        }
    }

    return Math.max(s.length - left, maxLen);
};

File: leetcode-java/159-longest-substring-with-at-most-two-distinct-characters.java
public class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        int a = 0;
        int b = -1;
        int max = 0;

        for (int i = 1; i < s.length(); i++) {
            if (s.charAt(i) == s.charAt(i - 1)) {
                continue;
            }

            if (b > -1 && s.charAt(i) != s.charAt(b)) {
                max = Math.max(max, i - a);
                a = b + 1;
            }

            b = i - 1;
        }

        return max > s.length() - a ? max : s.length() - a;
    }
}
